### DISCUSSION TITLE
Implementing "CFA for Target-Oriented Anomaly Localization"

### DISCUSSION THREAD

**Discussion by alexriedel1** [open-edge-platform/anomalib#388] [Feature Requests]
Hello,
I was just wondering if someone is already working on an anomalib implementation of https://arxiv.org/pdf/2206.04325.pdf
(https://github.com/sungwool/CFA_for_anomaly_localization)

It apparently scores higher than PatchCore on MVTec without any tricks like ensembling very large models.

**Comment by samet-akcay** [comment]
Hi @alexriedel1, the paper indeed looks promising. No one is working on it at the moment. It would be good to implement it for sure.

**Comment by aparico** [comment]
@samet-akcay Is there any update about implementing CFA yet?

**Reply by samet-akcay** [reply]
@aparico, it's on our to-do list, but we are a bit occupied with some other tasks unfortunately.

**Comment by StarShang** [comment]
How to understand the meaning of this code   in “src/anomalib/models/cfa/torch_model.py”：


```
def initialize_centroid(self, data_loader: DataLoader) -> None:

  if self.gamma_c > 1:
            # TODO: Create PyTorch KMeans class.
            k_means = KMeans(n_clusters=(self.scale[0] * self.scale[1]) // self.gamma_c, max_iter=3000)
            cluster_centers = k_means.fit(self.memory_bank.cpu()).cluster_centers_
            self.memory_bank = torch.tensor(cluster_centers, requires_grad=False).to(device)

```
**I didn't understand what he meant. Can you help to explain its role.** **TKS!**  @samet-akcay @alexriedel1

**Comment by samet-akcay** [comment]
Closing as this has already been implemented.
